### DISCUSSION TITLE
[Released] A/B Testing

### DISCUSSION THREAD

**Discussion by gsoverini** [vercel/community#84] [Idea]
Hi there!

Do you think Split Testing could be a valuable feature to implement into Vercel?
The company I work for is in the process of selecting a new hosting platform for our web application, and it's a duel between Netlify and Vercel.

I'm a big Vercel fan, but our marketers would die for having a feature like this: https://docs.netlify.com/site-deploys/split-testing/.

I'd like to know if this kind of feature is in your radar. Thank you!

**Comment by jdorn** [comment]
There currently isn't a great solution for split testing at the CDN level.  Netlify's feature is extremely primitive and only works for the most basic experiments.  A/B testing is possible on Cloudflare/Fastly using edge workers, but it can be very tedious to set up and maintain.

As an alternative for now, I might recommend running tests in code instead.  For static sites, you would serve the same HTML to all users and then modify the page on the client.  For SSR sites, you can run the split test on the server and avoid any "flashes", although that can make caching difficult.

There are a few libraries out there for this, but none of them are that great for React/NextJS apps.  I've been working on https://github.com/growthbook/growthbook-react to help with this use case.

**Reply by gsoverini** [reply]
> For SSR sites, you can run the split test on the server and avoid any "flashes", although that can make caching difficult.

Turning a static page into a SSR page just for the sake of running A/B tests is not worth it IMO. If a page can be static, it should be static.

> For static sites, you would serve the same HTML to all users and then modify the page on the client.

I think running A/B tests on the client side is a big limitation and that's why I was asking if Vercel had something on their roadmap to solve this. There are 2 possibilities for circumventing client-side testing on static pages:

1. Serving 2 different (static) pages, which is what Netlify would allow us to do.
2. Serving 1 static page and customize it through the "Edge-slice re-rendering", as Outsmartly does. Though, we didn't feel like adding another layer to our infrastructure at this moment and Outsmartly seems to be a start-up in its early stages.

If Vercel started to offer similar functionalities it would add great value to the entire platform.

**Reply by beadlespouse** [reply]
Does anyone from Vercel know if this is being worked on?

**Reply by leerob** [reply]
👋  We are working on an improved solution for A/B testing! Expect to hear more about this soon 😄

**Reply by matthewiiv** [reply]
@leerob and update on this? Moving over from Netlify to Vercel and very keen to know

**Reply by leerob** [reply]
https://github.com/vercel/next.js/discussions/29750

**Reply by leerob** [reply]
Expect more news soon!

**Reply by Jamiewarb** [reply]
That's super exciting

**Comment by leerob** [comment]
We've released official documentation now including solutions for handling A/B testing on Vercel: https://vercel.com/docs/concepts/solutions/ab-testing

**Comment by muhammad-saleh** [comment]
Just wanted to add that what Netlify’s have is great to be used as a way to release a change gradually.
Meaning that you can serve the customers multiple releases on prod and monitor them separately.
For example I can check on Sentry if a specific release has issues or not. This is different from A/B testing.
